### PR TITLE
Improve connection graph output

### LINKED ISSUE #800: Cluster graph command should distinguish between trusted and untrusted peers
**Describe the feature you are proposing**

The output of `ipfs-cluster-ctl health graph` should distinguish between which peers are in the trusted peerset and which are not.

### PR DESCRIPTION
**Pull request by kishansagathiya**
- Distinguish between trusted nodes (dark green) and untrusted nodes
(light green)
- Distinguish between Cluster nodes (filled) and IPFS nodes (hollow)
- Straight edge between Cluster node and it's corresponding IPFS node
- Cluster nodes always on top row, IPFS nodes always on bottom row
- Shorter labels for nodes

Fixes #800

### CODE REVIEW

File: cmd/ipfs-cluster-ctl/graph.go [hunk 1]
```diff
@@ -6,8 +6,8 @@ import (
 	"io"
 	"sort"
 
+	dot "github.com/kishansagathiya/go-dot"
```
Comment (kishansagathiya) [inline cmd/ipfs-cluster-ctl/graph.go]:
```suggestion
	dot "github.com/zenground0/go-dot"
```

after https://github.com/ZenGround0/go-dot/pull/4 gets in

**Comment by hsanjuan** [comment]
Per comment at https://github.com/ipfs/ipfs-cluster/issues/800#issuecomment-532171201, the main thing I don't like are the peer names. I think the labels should say: `IPFS: Qm*abc` (or similar). In the case of cluster, it would be interesting to use the Cluster `peername` when available. Other than that, it's looking really nice!

**Comment by hsanjuan** [comment]
@kishansagathiya can I get a new screenshot?

**Comment by kishansagathiya** [comment]
@hsanjuan 
Did you mean a sample graph generated from this change?

g

**Comment by kishansagathiya** [comment]
> Have we tried how it looks with square boxes rather than circles? Might look better.

I had tried it. It did not look good.

**Comment by hsanjuan** [comment]
> > Have we tried how it looks with square boxes rather than circles? Might look better.
> 
> I had tried it. It did not look good.

ok, no problem

**Comment by kishansagathiya** [comment]
> If the ipfs daemon is not running it should display something like a orange blob with an error label (right now it simplu does not show it)

I tried this. Orange does match with the color scheme that I picked. I have gone with a empty ellipse for disconnected IPFS peer (linked to its cluster peer with dotted line), and filled for connected IPFS daemon. I can fill disconnected IPFS peers with orange/red color, but I think keeping it empty send a more clear message that it's not there for there for now.

> The black font on the dark green cluster peer might look better as white (and bold?). Might we worth using a color scheme where all colors fit as well (lots of color schemes generators online)

I have picked one of the schemes provided by dot lang, brbg11. It is simple and serves our purpose. Just two colours with various intensity. Times Roman isn't very nice in bold, so changed to Ariel.

> The self node should have the same type of label as the rest (name+iD). Maybe the border of the box should be in a different color (if possible, otherwise the box itself).

There isn't any attribute for that. I have kept double borders for self peer.

g

**Comment by kishansagathiya** [comment]
One more example.
Light green is for untrusted
light brown is for swarm ipfs peer that are not linked to any cluster peer.

g

**Comment by hsanjuan** [comment]
IPFS peers which are not there need to appear as error boxes because they should be there..

**Comment by kishansagathiya** [comment]
Error IPFS nodes in box shape
g

**Comment by hsanjuan** [comment]
Then make all ipfs daemons boxes?

**Comment by hsanjuan** [comment]
I've mentioned this to @ericronne, see if someone from design has some ideas on how things look best.

File: cmd/ipfs-cluster-ctl/graph.go [hunk 2]
```diff
@@ -78,53 +84,138 @@ type dotWriter struct {
...
+		node.FontColor = "6"
 		dW.clusterNodes[id] = &node
-	case tIpfs:
+	case tIPFS:
+		node.ID = fmt.Sprintf("I%d", len(dW.ipfsNodes))
+		node.Shape = "box"
+		ipfsID, ok := dW.clusterIpfsEdges[id]
+		if !ok {
+			node.Label = label("IPFS", "Errored")
+			node.ColorScheme = "X11"
```
Comment (hsanjuan) [inline cmd/ipfs-cluster-ctl/graph.go]: How about limiting to colors in the `rdylbu9` color scheme

File: cmd/ipfs-cluster-ctl/graph.go [hunk 3]
```diff
@@ -31,8 +31,11 @@ import (
 type nodeType int
 
 const (
-	tCluster nodeType = iota // The cluster node type
-	tIpfs                    // The IPFS node type
+	tSelfCluster    nodeType = iota //  cluster self node
+	tCluster                        //  cluster node
+	tTrustedCluster                 //  trusted cluster node
+	tIPFS                           //  IPFS node linked to a Cluster node
+	tSwarmIPFS                      //  IPFS node linked to a tIPFS node
```
Comment (hsanjuan) [inline cmd/ipfs-cluster-ctl/graph.go]: What are tSwarmPeers vs tIPFS ? Am I understanding it right that you are printing nodes in the IPFS swarm that are not connected to any cluster peer? Does this mean that if I run connect graph on an IPFS peer connected to 400 other ipfs peers in the network I am going to get 400 boxes? If so, please remove. We're only interested in ipfs peers attached to cluster peers.

File: cmd/ipfs-cluster-ctl/graph.go (hunk 2 above)
Comment (hsanjuan) [reply cmd/ipfs-cluster-ctl/graph.go]: (and only using colors from that scheme, for all things)

File: cmd/ipfs-cluster-ctl/graph.go [hunk 4]
```diff
@@ -78,53 +84,138 @@ type dotWriter struct {
...
 	default:
 		return errUnknownNodeType
 	}
-	dW.dotGraph.AddVertex(&node)
 
+	graph.AddVertex(&node)
 	return nil
 }
 
+func shorten(id string) string {
```
[the remark this answers is not in the document]
Comment (kishansagathiya) [reply cmd/ipfs-cluster-ctl/graph.go]: that would return something like `<peer.ID p>`
It makes the string unnecessarily long, making the graph bigger and uglier.
I can how make the actual shorten peer id consistent by using `pid[:2] + "*" +pid[len(pid)-6:]`

[the remark this answers is not in the document]
Comment (hsanjuan) [reply cmd/ipfs-cluster-ctl/graph.go]: yes

File: cmd/ipfs-cluster-ctl/graph.go (hunk 2 above)
Comment (kishansagathiya) [reply cmd/ipfs-cluster-ctl/graph.go]: It doesn't have white color. We would have to rely on light yellow or light blue for for fonts in darker ellipse/boxes, which I think would not be as good as white.

**Comment by kishansagathiya** [comment]
I am leaving test modification for later.

**Comment by hsanjuan** [comment]
When I test this I don't see the connections between the ipfs daemons in the cluster. I think you removed too much in your last commit. Also, let's make boxes for the cluster peers. Boxes looks better

**Comment by kishansagathiya** [comment]
> When I test this I don't see the connections between the ipfs daemons in the cluster. I think you removed too much in your last commit.

I tried twice just now, and I see them connected.
g

Just a guess, did you modify your godot for testing and left it changed?

**Comment by hsanjuan** [comment]
I think this is nice enough:


image

Let's move on on this

**Comment by hsanjuan** [comment]
@kishansagathiya it seems tests failed randomly depending on the order of the map. I fixed that on other branch.

**Pull request merged**